src/index.js

```diff
@@ -1,39 +1,51 @@
 import React from 'react'
 import PropTypes from 'prop-types'
 
-export const macro = (template, opts = {}) => {
+export const macro = (components = {}, template) => {
+  const componentKeys = Object.keys(components)
+
   class Macro extends React.Component {
     static propTypes = {
-      children: Array.isArray(opts.childTypes)
-        ? (props, name) => {
-          const children = React.Children.toArray(props.children)
-          for (let i = 0; i < children.length; i++) {
-            const child = children[i]
-            if (opts.childTypes.includes(child.type)) continue
-            return new Error(
-              'Invalid child component `' + child.type + '`'
-            )
-          }
+      children: (props, name) => {
+        const children = React.Children.toArray(props.children)
+        for (let i = 0; i < children.length; i++) {
+          const child = children[i]
+
+          if (componentKeys.includes(child.type.macroName)) continue
+
+          return new Error(
+            [
+              'Invalid child component `',
+              child.type,
+              '`. ',
+              'Must be one of: ',
+              componentKeys.join(', ')
+            ].join('')
+          )
         }
-        : PropTypes.node
+      }
     }
 
     constructor (props) {
       super()
 
-      this.getElements = children =>
+      this.parseChildren = children =>
         React.Children.toArray(children)
-          .map(child => ({
-            key: getName(child),
-            type: child.type,
-            element: child
-          }))
-          .reduce((a, b) => ({
+          .reduce((a, child) => ({
```

Comment:
Nit: if this is a hot code path it is better to use mutation than spreads+reduce